File: android/app/src/main/java/com/xirigo/ecommerce/core/designsystem/component/XGWishlistButton.kt
package com.xirigo.ecommerce.core.designsystem.component

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.Animatable
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Favorite
import androidx.compose.material.icons.outlined.FavoriteBorder
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.IconButtonDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.xirigo.ecommerce.R
import com.xirigo.ecommerce.core.designsystem.theme.XGColors
import com.xirigo.ecommerce.core.designsystem.theme.XGMotion
import com.xirigo.ecommerce.core.designsystem.theme.XGTheme

// Token source: components/atoms/xg-wishlist-button.json
private val ButtonSize = 32.dp
private val ButtonIconSize = 16.dp
private val ButtonCornerRadius = 16.dp
private val ButtonElevation = 2.dp

/** Peak scale for the spring bounce effect on toggle. */
private const val BOUNCE_SCALE = 1.2f

/**
 * Toggle button for wishlist state with heart icon.
 *
 * Motion tokens (from `XGMotion`):
 * - Color transition: `XGMotion.Easing.standardTween(XGMotion.Duration.INSTANT)` (100ms)
 * - Scale bounce: `XGMotion.Easing.springSpec()` (dampingRatio=0.7, stiffness=Medium)
 */
@Composable
fun XGWishlistButton(
    isWishlisted: Boolean,
    onToggle: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val contentDescription = if (isWishlisted) {
        stringResource(R.string.common_remove_from_wishlist)
    } else {
        stringResource(R.string.common_add_to_wishlist)
    }

    val icon = if (isWishlisted) Icons.Filled.Favorite else Icons.Outlined.FavoriteBorder

    val animatedTint by animateColorAsState(
        targetValue = if (isWishlisted) XGColors.BrandPrimary else XGColors.OnSurfaceVariant,
        animationSpec = XGMotion.Easing.standardTween(XGMotion.Duration.INSTANT),
        label = "wishlistTint",
    )

    val scaleAnimatable = remember { Animatable(1f) }
    LaunchedEffect(isWishlisted) {
        scaleAnimatable.snapTo(BOUNCE_SCALE)
        scaleAnimatable.animateTo(
            targetValue = 1f,
            animationSpec = XGMotion.Easing.springSpec(),
        )
    }

    val shape = RoundedCornerShape(ButtonCornerRadius)

    IconButton(
        onClick = onToggle,
        modifier = modifier
            .size(ButtonSize)
            .shadow(elevation = ButtonElevation, shape = shape)
            .clip(shape)
            .scale(scaleAnimatable.value),
        colors = IconButtonDefaults.iconButtonColors(
            containerColor = XGColors.Surface,
        ),
    ) {
        Icon(
            imageVector = icon,
            contentDescription = contentDescription,
            modifier = Modifier.size(ButtonIconSize),
            tint = animatedTint,
        )
    }
}

@Preview(showBackground = true)
@Composable
private fun XGWishlistButtonInactivePreview() {
    XGTheme {
        XGWishlistButton(isWishlisted = false, onToggle = {})
    }
}

@Preview(showBackground = true)
@Composable
private fun XGWishlistButtonActivePreview() {
    XGTheme {
        XGWishlistButton(isWishlisted = true, onToggle = {})
    }
}

File: ios/XiriGoEcommerce/Core/DesignSystem/Component/XGWishlistButton.swift
import SwiftUI

// MARK: - XGWishlistButton

/// A heart icon toggle button overlaid on product cards.
/// Token source: `components/atoms/xg-wishlist-button.json`.
///
/// - Size: 32x32pt, circular
/// - Active: filled heart in brand primary
/// - Inactive: outlined heart in text secondary
///
/// **Motion tokens** (from `XGMotion`):
/// - Color transition: `.easeInOut(duration: XGMotion.Duration.instant)` (0.1s)
/// - Scale bounce: `XGMotion.Easing.spring` (response=0.35, dampingFraction=0.7)
struct XGWishlistButton: View {
    // MARK: - Lifecycle

    init(
        isWishlisted: Bool,
        onToggle: @escaping () -> Void,
    ) {
        self.isWishlisted = isWishlisted
        self.onToggle = onToggle
    }

    // MARK: - Internal

    var body: some View {
        Button(
            action: {
                bounceScale = Constants.bounceScale
                withAnimation(.easeInOut(duration: XGMotion.Duration.instant)) {
                    onToggle()
                }
                withAnimation(XGMotion.Easing.spring) {
                    bounceScale = 1.0
                }
            },
            label: {
                Image(systemName: isWishlisted ? "heart.fill" : "heart")
                    .font(.system(size: Constants.iconSize))
                    .foregroundStyle(
                        isWishlisted
                            ? XGColors.brandPrimary
                            : XGColors.onSurfaceVariant,
                    )
                    .frame(
                        width: Constants.buttonSize,
                        height: Constants.buttonSize,
                    )
                    .background(XGColors.surface)
                    .clipShape(Circle())
                    .xgElevation(XGElevation.level2)
                    .scaleEffect(bounceScale)
                    .accessibilityHidden(true)
            },
        )
        .frame(minWidth: XGSpacing.minTouchTarget, minHeight: XGSpacing.minTouchTarget)
        .accessibilityLabel(
            isWishlisted
                ? String(localized: "common_remove_from_wishlist")
                : String(localized: "common_add_to_wishlist"),
        )
        .accessibilityAddTraits(.isToggle)
        .animation(.easeInOut(duration: XGMotion.Duration.instant), value: isWishlisted)
    }

    // MARK: - Private

    private enum Constants {
        static let buttonSize: CGFloat = 32
        static let iconSize: CGFloat = 16
        static let bounceScale: CGFloat = 1.2
    }

    @State private var bounceScale: CGFloat = 1.0

    private let isWishlisted: Bool
    private let onToggle: () -> Void
}

// MARK: - Previews

#Preview("XGWishlistButton") {
    HStack(spacing: XGSpacing.lg) {
        XGWishlistButton(isWishlisted: false, onToggle: {})
        XGWishlistButton(isWishlisted: true, onToggle: {})
    }
    .padding()
    .xgTheme()
}
